[rust-sdk] docs and examples

stefan-mysten opened the pull request:
## Description 

First step in improving the usability of the Sui Rust SDK. This PR adds doc comments and examples that can be run on a local Sui network. Polish will still be required but before that I'd like to get some more eyes on this to see if we're on the right track. There are some lacking examples and some missing doc comments, which I will address in a follow-up PR where I plan to also implement more robust testing.

* See the docs
`cd crates/sui-sdk && cargo doc --open`

* Run the examples
  ```bash
  cargo run --example sui_client
  cargo run --example coin_read_api
  cargo run --example event_api
  cargo run --example governance_api
  cargo run --example read_api
  cargo run --example transaction_builder_api
  ```
## Test Plan 

Same tests as before. 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
Add Rust SDK docs and examples.

bmwill reviewed (changes requested):
Taken a look through most of this and I really appreciate the extra doc comments that have been added but not all of the rust code examples add a whole lot of value over just reading the text documentation and the function signatures. In addition to that, in order to fix CI you'll likely want to mark many of the examples with `no_run` so that they don't get run as doc tests.

stefan-mysten commented on crates/sui-sdk/examples/utils/mod.rs:
```diff
@@ -0,0 +1,45 @@
+}
+
+pub(crate) async fn request_tokens_from_faucet(address: SuiAddress) -> Result<(), anyhow::Error> {
+    let mut map = HashMap::new();
+    let mut recipient = HashMap::new();
+    recipient.insert("recipient", address.to_string());
+    map.insert("FixedAmountRequest", recipient);
```
Perhaps I'm missing something but can't figure out how to do this. 
`sui-faucet` imports `sui-sdk`, so importing `sui-faucet` in the sdk to use `FaucetRequest` type leads to a dependency cycle.

stefan-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,142 @@
+
+## Getting Started
+
+We are currently working on publishing the Sui Rust SDK crate on crates.io. Until then, add the `sui-sdk` dependency as following:
```
Fixed

stefan-mysten commented on crates/sui-sdk/README.md:
````diff
@@ -1,6 +1,142 @@
+<!-- 
+```toml
+sui_sdk = "0.1"
+``` -->
````
Fixed

stefan-mysten commented on crates/sui-sdk/examples/coin_read_api.rs:
```diff
@@ -0,0 +1,100 @@
+
+mod utils;
+use sui_sdk::SuiClientBuilder;
+use utils::sui_address_for_examples;
```
This should be better now, let me know otherwise.

stefan-mysten commented on crates/sui-sdk/examples/coin_read_api.rs:
```diff
@@ -0,0 +1,100 @@
+
+mod utils;
+use sui_sdk::SuiClientBuilder;
+use utils::sui_address_for_examples;
```
This should be better now, let me know otherwise.

stefan-mysten commented on crates/sui-sdk/examples/transaction_builder_api.rs:
```diff
@@ -0,0 +1,84 @@
+    let keystore = Keystore::InMem(InMemKeystore::new_insecure_for_tests(2));
+    let addresses = keystore.addresses();
+    let Some(sender) = addresses.get(0) else {panic!("Keystore did not manage to generate two keys in memory. Aborting")};
+    let Some(recipient) = addresses.get(1) else {panic!("Keystore did not manage to generate two keys in memory. Aborting")};
```
Fixed

stefan-mysten commented on crates/sui-sdk/examples/transfer_coins.rs:
```diff
@@ -2,41 +2,43 @@
+    let keystore = Keystore::InMem(InMemKeystore::new_insecure_for_tests(2));
+    let addresses = keystore.addresses();
+    let Some(sender_address) = addresses.get(0) else {panic!("Keystore did not manage to generate two keys in memory. Aborting")};
+    let Some(recipient_address) = addresses.get(1) else {panic!("Keystore did not manage to generate two keys in memory. Aborting")};
```
Fixed

stefan-mysten commented on crates/sui-sdk/src/apis.rs:
````diff
@@ -47,7 +48,19 @@ impl ReadApi {
+    /// let sui = SuiClientBuilder::default().build_localnet().await?; // local Sui network
+    /// let address = SuiAddress::from_str("sui_address_here")?;
+    /// let owned_objects = sui.read_api().get_owned_objects(address, None, None, None).await?;
+    /// ```
````
Added the `no_run` marker to all examples.

stefan-mysten commented on crates/sui-sdk/src/apis.rs:
````diff
@@ -319,7 +503,19 @@ impl CoinReadApi {
+    /// let sui = SuiClientBuilder::default().build_localnet().await?; // local Sui network
+    /// let address = SuiAddress::from_str("sui_address_here")?;
+    /// let coins = sui.coin_read_api().get_all_coins(address, None, None).await?;
+    /// ```
````
I agree with you. I struggled with finding the right balance. Perhaps we should remove all these kind of examples, and instead point to the `examples` folder.

ronny-mysten commented on crates/sui-sdk/src/apis.rs:
```diff
@@ -306,6 +477,19 @@ impl CoinReadApi {
+    /// use std::str::FromStr;
+    ///
+    /// let sui = SuiClientBuilder::default().build_localnet().await?; // local Sui network
+    /// let address = SuiAddress::from_str("sui_address_here")?;
```
We typically write with angle brackets and all caps, as in `<SENDER_ADDRESS>`. This provides a consistent scheme regardless of data type and tells the user what's expected there. We can certainly change that approach, but it needs to be consistent across docs.

stefan-mysten replied:
@ronny-mysten You mean to write like this `SuiAddress::from_str(<SENDER_ADDRESS>)?;` ?

ronny-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+# Sui Rust SDK
 
-This directory contains examples of interacting with a Move language smart contract using the Sui Rust SDK. See the [introduction to the Rust SDK](https://github.com/MystenLabs/sui/blob/main/doc/src/build/rust-sdk.md) for additional details.
+This crate provides the Sui Rust SDK. For the crate documentation, see [https://docs.rs/sui_sdk/](https://docs.rs/sui_sdk/)
```
```suggestion
This crate provides the Sui Rust SDK. For the crate documentation, see [https://docs.rs/sui_sdk/](https://docs.rs/sui_sdk/).
```

bmwill replied:
The crate isn't published on crates.io so we should remove the broken link and not link there yet

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
-This directory contains examples of interacting with a Move language smart contract using the Sui Rust SDK. See the [introduction to the Rust SDK](https://github.com/MystenLabs/sui/blob/main/doc/src/build/rust-sdk.md) for additional details.
+This crate provides the Sui Rust SDK. For the crate documentation, see [https://docs.rs/sui_sdk/](https://docs.rs/sui_sdk/)
+
+## Getting Started
```
```suggestion
## Getting started
```
Title is title case, but following headings are sentence case.

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+
+The main building block for the Sui Rust SDK is the `SuiClientBuilder`, which provides a simple and straightforward way of connectiong to a Sui network and having access to the different available APIs. 
+
+A simple example that connects to a running Sui local network and available Sui networks is shown below. If you want to try to run this program, make sure to spin up a local network with a local validator, a fullnode, and a faucet server (see [preqrequisites](README.md) for more inforrmation).
```
```suggestion
An example that connects to a running Sui local network and available Sui networks follows. If you want to run this program, make sure to spin up a local network with a local validator, a Full node, and a faucet server (see [preqrequisites](README.md) for more inforrmation).
```
Be careful with words like 'simple'. If it's not simple for the reader for whatever reason, you're unintentionally calling them dumb :)

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
````diff
@@ -1,6 +1,138 @@
+
+```
+
+# Rust SDK Examples
````
```suggestion
# Rust SDK examples
```

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+
+# Rust SDK Examples
+
+The [examples](https://github.com/MystenLabs/sui/tree/main/crates/sui-sdk/examples) folder provides both simple and advanced examples.
```
```suggestion
The [examples](https://github.com/MystenLabs/sui/tree/main/crates/sui-sdk/examples) folder provides both basic and advanced examples.
```

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+Unless otherwise specified, most of these examples assume that Sui is installed, and that there is a local network running.
+
+* Install `sui` binary following the [Sui installation](https://github.com/MystenLabs/sui/blob/main/doc/src/build/install.md##install-sui-binaries) docs.
+* For local development, after the `sui` binary is installed, run `sui-test-validator` (or `cargo run --bin sui-test-validator` in the [cloned Git repository](https://github.com/mystenlabs/sui) if you don't want to install `Sui`) to spin up a local network with a local validator, a fullnode, and a faucet server. Refer to [this guide](https://docs.sui.io/build/sui-local-network) for more informati ...
```
```suggestion
* For local development, after you install the `sui` binary, run `sui-test-validator` (or `cargo run --bin sui-test-validator` in the [cloned Git repository](https://github.com/mystenlabs/sui) if you don't want to install `Sui`) to spin up a local network with a local validator, a Full node, and a faucet server. Refer to [this guide](https://docs.sui.io/build/sui-local-network) for more information. The local network will be up and running on `http://127.0.0.1:9000` and the faucet server on `http://127.0.0.1:9123`. 
```

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+## Basic Examples
+
+### Connecting to Sui Network
+The `SuiClientBuilder` struct provides a connection to the JSON-RPC Server and should be used for all read-only operations. The default URLs to connect to the Sui network are:
```
```suggestion
The `SuiClientBuilder` struct provides a connection to the JSON-RPC server that you use for all read-only operations. The default URLs to connect to the Sui network are:
```

stefan-mysten replied:
Fixed

ronny-mysten commented on crates/sui-sdk/README.md:
````diff
@@ -1,6 +1,138 @@
+}
+```
+
+## Advanced Examples
````
```suggestion
## Advanced examples
```

stefan-mysten replied:
Fixed

ronny-mysten reviewed (commented):
Awesome, Stefan. Many thanks! I noted some style guide nits but this will be helpful.

stefan-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+See the transaction builder [example](examples/transaction_builder_api.rs).
+
+
+## Games Examples
```
Fixed

stefan-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,138 @@
+
+## Running the examples
+
+In the root folder of the `sui` repository (or in the `sui-sdk` crate folder), examples can be individually run using the command  `cargo run --example filename` (without `.rs` extension). For example:
```
Fixed

stefan-mysten commented:
@all - thanks for your valuable feedback, it was great having other people scrutinizing the code!
@manolisliolios - let me know if you manage to try it out. 
@ronny-mysten - I think I addressed all your comments. Let me know if there's something else you think we should do.
@bmwill - not sure how to handle the FaucetRequest change, so let me know if you have suggestions.

amnn commented on crates/sui-sdk/examples/transaction_builder_api.rs:
```diff
@@ -0,0 +1,90 @@
+    let Some(coin) = sui
+        .coin_read_api()
+        .get_coins(*sender, None, None, None)
+        .await?.next_cursor else {panic!("Faucet did not work correctly and the provided Sui address has no coins")};
```
nit: formatting (`cargo fmt` is not aware of `let ... else` syntax so you need to help it out with that).

amnn commented on crates/sui-sdk/src/apis.rs:
````diff
@@ -47,7 +48,19 @@ impl ReadApi {
+    ///
+    /// # Examples
+    ///
+    /// ```rust, ignore
````
Can we use `no_run` instead of `ignore` so we at least compile these snippets?

amnn commented on crates/sui-sdk/src/lib.rs:
```diff
@@ -197,25 +394,30 @@ impl Debug for RpcClient {
 }
 
 impl SuiClient {
+    /// Returns a list of available RPC calls
     pub fn available_rpc_methods(&self) -> &Vec<String> {
         &self.api.info.rpc_methods
     }
 
+    /// Returns a list of available subscriptions
     pub fn available_subscriptions(&self) -> &Vec<String> {
         &self.api.info.subscriptions
     }
 
+    /// Returns the API version information
```
These doc comments seem a bit redundant on their own.  Here's some information that might be more helpful here:

- What is an RPC Method vs a Subscription
- Why do these values change? (i.e. why would a given RPC method be available in some cases but not others).
- What information is included in "API version information" other than the version itself?

amnn commented on crates/sui-sdk/src/lib.rs:
```diff
@@ -227,26 +429,35 @@ impl SuiClient {
-    }
+    /// Returns a reference to the coin read API
     pub fn coin_read_api(&self) -> &CoinReadApi {
         &self.coin_read_api
     }
+
+    /// Returns a reference to the event API
     pub fn event_api(&self) -> &EventApi {
         &self.event_api
     }
+
+    /// Returns a reference to the governance API
+    pub fn governance_api(&self) -> &GovernanceApi {
+        &self.governance_api
+    }
+
+    /// Returns a reference to the quorum driver API
     pub fn quorum_driver_api(&self) -> &QuorumDriverApi {
         &self.quorum_driver_api
     }
-    pub fn governance_api(&self) -> &GovernanceApi {
-        &self.governance_api
+
+    /// Returns a reference to the read API
+    pub fn read_api(&self) -> &ReadApi {
+        &self.read_api
+    }
+
+    /// Returns a reference to the transaction builder API
+    pub fn transaction_builder(&self) -> &TransactionBuilder {
+        &self.transaction_builder
     }
```
Similar comment here -- better not to have a doc comment at all than to have one that just repeats the function name.  But in these cases, maybe you can add doc comments explaining what you would use these APIs for?

amnn commented on crates/sui-sdk/examples/coin_read_api.rs:
These examples seem a little basic to me -- I'm not sure if they add much value over the examples that you give in the doc comments of the various APIs.  What might be more interesting is to see the APIs being combined in some common ways.

amnn commented on crates/sui-sdk/examples/governance_api.rs:
```diff
@@ -0,0 +1,54 @@
+    let committee = sui_local.governance_api().get_committee_info(None).await?; // None defaults to the last epoch
+
+    println!(" *** Committee Info ***");
+    println!("{:?}", committee);
```
Building on my comment on `coin_api.rs` -- rather than just printing the output, how about, e.g. combining these APIs so that you can display the validator details for the coins you've staked, or maybe you calculate your portion of that validator's stake, in basis points?

stefan-mysten replied:
Great idea, will do it!

amnn reviewed (commented):
Thanks for working on this @stefan-mysten !  High level comments from me:

- Can we use `no_run` instead of `ignore` so that we at least compile the examples?
- We should avoid perfunctory doc comments that just rephrase the function name in English.  If that's all we can say about the function, then I think we should be fine to just not have a doc comment.  There's a couple of these examples in the PR, but I think in all those cases, we can say something more about the function.
- I think the `examples` can be a little bit more involved -- find ways to combine APIs in interesting, practical ways, because that will add more value than calling each API in turn.

stefan-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,139 @@
+Unless otherwise specified, most of these examples assume that Sui is installed, and that there is a local network running.
+
+* Install `sui` binary following the [Sui installation](https://github.com/MystenLabs/sui/blob/main/doc/src/build/install.md##install-sui-binaries) docs.
+* For local development, after you install the `sui` binary, run `sui-test-validator` (or `cargo run --bin sui-test-validator` in the [cloned Git repository](https://github.com/mystenlabs/sui) if you don't want to install `Sui`) to spin up a local network with a local validator, a Full node, and a faucet server. Refer to [this guide](https://docs.sui.io/build/sui-local-network) for more informati ...
```
Fixed

stefan-mysten commented on crates/sui-sdk/README.md:
````diff
@@ -1,6 +1,139 @@
+
+An example that connects to a running Sui local network and available Sui networks follows. If you want to run this program, make sure to spin up a local network with a local validator, a Full node, and a faucet server (see [preqrequisites](README.md) for more inforrmation).
+
+```rust, ignore
````
Fixed

stefan-mysten commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,139 @@
+
+The main building block for the Sui Rust SDK is the `SuiClientBuilder`, which provides a simple and straightforward way of connectiong to a Sui network and having access to the different available APIs. 
+
+An example that connects to a running Sui local network and available Sui networks follows. If you want to run this program, make sure to spin up a local network with a local validator, a Full node, and a faucet server (see [preqrequisites](README.md) for more inforrmation).
```
Fixed

amnn commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,134 @@
+
+There are five files ending in `_api.rs` which provide code examples of the corresponding APIs and their methods. These showcase how to use the Sui Rust SDK, and can run against the Sui testnet. Below are instructions on the prerequisites and how to run these examples.
+
+## Preqrequisites
```
nit, sp.

amnn commented on crates/sui-sdk/README.md:
````diff
@@ -1,6 +1,134 @@
+
+
+```rust
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
````
nit: remove the license in the code snippet? You have one for the README file overall.

amnn commented on crates/sui-sdk/examples/coin_read_api.rs:
```diff
@@ -0,0 +1,93 @@
+    // Get coins
+    let coins = sui
+        .coin_read_api()
+        .get_coins(active_address, None, None, Some(5)) // get the first five coins. Note that this can be filtered by coin type: coin_type: Some("0x2::sui::SUI".to_string())
```
nit: I'm not sure whether `cargo fmt` will help with this, but can we limit the line length to either 80 or 100 characters? (Whatever is the limit we use throughout the rest of the codebase -- I think it's 80).

amnn commented on crates/sui-sdk/examples/event_api.rs:
```diff
@@ -0,0 +1,42 @@
+    let sui = SuiClientBuilder::default().build_testnet().await?; // testnet Sui network
+    println!("Sui testnet version{:?}", sui.api_version());
+
+    // TODO - make this work
```
Is this going to be done as an immediate follow-up? If not, let's remove the comment, othewise we're just going to confuse people with a reference that doesn't work.

amnn commented on crates/sui-sdk/examples/event_api.rs:
```diff
@@ -0,0 +1,42 @@
+    println!("{:?}", events);
+    println!(" *** Get events ***\n ");
+
+    let query_events = sui
+        .event_api()
+        .query_events(EventFilter::All(vec![]), None, Some(5), true) // query first 5 events in descending order
+        .await?;
```
nit: use a pattern that allows you to name the constant:
```suggestion
    let descending = true;
    let query_events = sui
        .event_api()
        .query_events(EventFilter::All(vec![]), None, Some(5), descending) // query first 5 events in descending order
        .await?;
```

amnn commented on crates/sui-sdk/examples/transaction_builder_api.rs:
```diff
@@ -0,0 +1,91 @@
+    let txb_res = txb
+        .transfer_object(*sender, coin, None, 5000000, *recipient)
+        .await?;
+    println!("Tranfer object data: {:?}", txb_res);
```
nit sp.

```suggestion
    println!("Transfer object data: {:?}", txb_res);
```

amnn commented on crates/sui-sdk/src/apis.rs:
```diff
@@ -47,7 +48,23 @@ impl ReadApi {
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), anyhow::Error> {
+    ///     let sui = SuiClientBuilder::default().build_localnet().await?; // local Sui network
```
I vaguely recall that it's possible to hide set-up code (`use` statements, the `main` function boilerplate, etc) to limit noise in the doc comments -- have a look at the output you get from `cargo doc` but that may be something worth considering for these.

stefan-mysten replied:
We can do that for sure, but I think it might be useful to have them nevertheless. Tokio seems to have most examples like that: https://docs.rs/tokio/1.29.1/tokio/net/fn.lookup_host.html.

amnn replied:
Fair, your call!  I think tokio is a little bit of a special case example though because it "takes over" the implementation of main, so it's relevant to include that in their docs (but you could also make an argument that in our case, it's worth knowing that we use tokio).

amnn reviewed (commented):
I would consider splitting this PR up into (at least) two pieces:

 - Adding the doc comments, which seems already good to go.
 - Adding `examples`, which you can split each into their own PR, and there's some more comments for.

Otherwise it becomes unclear on each PR update what still needs to be reviewed.

stefan-mysten commented:
Thanks all for your help. I am splitting this PR into two PRs, one for doc comments, and one for examples! 
Superseeded by #13045

amnn commented on crates/sui-sdk/README.md:
```diff
@@ -1,6 +1,139 @@
+### Connecting to Sui Network
+The `SuiClientBuilder` struct provides a connection to the JSON-RPC server that you use for all read-only operations. The default URLs to connect to the Sui network are:
+
+- Local: http://127.0.0.1:9000
```
That tracks, thanks for checking on that!
